source/s3_client.c

```diff
@@ -293,17 +346,18 @@ struct aws_s3_client *aws_s3_client_new(
     /* Make a copy of the region string. */
     client->region = aws_string_new_from_array(allocator, client_config->region.ptr, client_config->region.len);
 
-    if (client_config->part_size != 0) {
-        *((size_t *)&client->part_size) = (size_t)client_config->part_size;
-    } else {
-        *((size_t *)&client->part_size) = s_default_part_size;
-    }
+    *((size_t *)&client->part_size) = part_size;
 
     if (client_config->max_part_size != 0) {
         *((uint64_t *)&client->max_part_size) = client_config->max_part_size;
     } else {
         *((uint64_t *)&client->max_part_size) = s_default_max_part_size;
     }
+
+    if (client_config->max_part_size > s_default_max_part_size_based_on_mem_limit(mem_limit)) {
+        *((uint64_t *)&client->max_part_size) = s_default_max_part_size_based_on_mem_limit(mem_limit);
```

Comment:
this could be a breaking change.

It's unlikely that anyone was setting config->max_part_size, so they were allowed to upload 4GiB (32bit system) or 5GiB (64bit system) parts by default.

But now, the largest they can upload is 1/4 the default memory-limit, so 256MiB or 512MiB on 32bit, or 2GiB 1GiB or 512MiB on 64bit

Reply:
its max part size. this will switch max part size from 4GB to 512MB and try to enforce 2GB in flight. I doubt if anything close to 4GB part size ever worked in the past unless we were able to enforce that only a couple parts were running at the same time, given that we panic on failed allocations.